flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputSelectableProcessor.java

```diff
@@ -286,64 +316,33 @@ private void checkAndSetAvailable(int inputIndex) {
 		}
 	}
 
-	/**
-	 * @return false if both of the inputs are finished, true otherwise.
-	 */
-	private boolean waitForAvailableInput(InputSelection inputSelection)
-		throws ExecutionException, InterruptedException, IOException {
-
+	private CompletableFuture<?> waitForAvailableInput(InputSelection inputSelection) {
 		if (inputSelection.isALLMaskOf2()) {
 			return waitForAvailableEitherInput();
 		} else {
-			waitForOneInput(
+			return waitForOneInput(
 				(inputSelection.getInputMask() == InputSelection.FIRST.getInputMask()) ? input1 : input2);
-			return true;
 		}
 	}
 
-	private boolean waitForAvailableEitherInput()
-		throws ExecutionException, InterruptedException {
-
-		CompletableFuture<?> future1 = input1.isFinished() ? UNAVAILABLE : input1.isAvailable();
-		CompletableFuture<?> future2 = input2.isFinished() ? UNAVAILABLE : input2.isAvailable();
-
-		if (future1 == UNAVAILABLE && future2 == UNAVAILABLE) {
-			return false;
+	private CompletableFuture<?> waitForAvailableEitherInput() {
+		if (input1.isFinished() && input2.isFinished()) {
+			return AVAILABLE;
 		}
-
-		// block to wait for a available input
-		CompletableFuture.anyOf(future1, future2).get();
-
-		if (future1.isDone()) {
-			setAvailableInput(input1.getInputIndex());
+		if (input1.isFinished()) {
+			return input2.isAvailable();
 		}
-		if (future2.isDone()) {
-			setAvailableInput(input2.getInputIndex());
+		if (input2.isFinished()) {
+			return input1.isAvailable();
 		}
-
-		return true;
+		return CompletableFuture.anyOf(input1.isAvailable(), input2.isAvailable());
 	}
 
-	private void waitForOneInput(StreamTaskInput input)
-		throws IOException, ExecutionException, InterruptedException {
-
+	private CompletableFuture<?> waitForOneInput(StreamTaskInput input) {
 		if (input.isFinished()) {
-			throw new IOException("Could not read the finished input: input" + (input.getInputIndex() + 1) +  ".");
+			throw new IllegalStateException("Could not read the finished input: input" + (input.getInputIndex() + 1) +  ".");
```

Comment:
Personally, I'd inline this method.
Otherwise:
 * `Preconditions.checkState`? 😃
 * Same here, suggestion to give it a new name.

Reply:
I have also realised, that this shouldn't be a `IllegalStateException`, since this is a user error - I have refactored the code accordingly.